src/QMCWaveFunctions/EinsplineSetBuilderCommon.cpp

```diff
@@ -398,101 +399,139 @@ bool EinsplineSetBuilder::TwistPair(PosType a, PosType b)
+      PosType superTwist = dot(S, primTwist);
+      PosType kp         = PrimCell.k_cart(primTwist);
+      PosType ks         = SuperCell.k_cart(superTwist);
+      if (dot(ks - kp, ks - kp) > 1.0e-6)
+      {
+        app_error() << "Primitive and super k-points do not agree.  Error in coding.\n";
+        APP_ABORT("EinsplineSetBuilder::AnalyzeTwists2");
+      }
+      PosType frac = FracPart(superTwist);
+      bool found   = false;
+      for (int j = 0; j < superFracs.size(); j++)
+      {
+        PosType diff = frac - superFracs[j];
+        if (dot(diff, diff) < 1.0e-6)
+        {
+          found = true;
+          superIndex.push_back(j);
+        }
+      }
+      if (!found)
+      {
+        superIndex.push_back(superFracs.size());
+        superFracs.push_back(frac);
+      }
     }
-    PosType frac = FracPart(superTwist);
-    bool found   = false;
-    for (int j = 0; j < superFracs.size(); j++)
+    const int numSuperTwists = superFracs.size();
+    app_log() << "Found " << numSuperTwists << " distinct supercell twists.\n";
+    // For each supercell twist, create a list of primitive twists which
+    // belong to it.
+    superSets.resize(numSuperTwists);
+    for (int ki = 0; ki < numPrimTwists; ki++)
+      superSets[superIndex[ki]].push_back(ki);
+    app_log() << "number of things" << std::endl;
+    app_log() << TwistSymmetry.size() << std::endl;
+    app_log() << TwistWeight.size() << std::endl;
+    //     for (int ki=0; ki<TwistSymmetry.size(); ki++)
+    //       fprintf (stderr, "%d %d %d\n",ki,TwistSymmetry[ki],TwistWeight[ki]);
+    if (myComm->rank() == 0)
     {
-      PosType diff = frac - superFracs[j];
-      if (dot(diff, diff) < 1.0e-6)
+      int n_tot_irred(0);
+      for (int si = 0; si < numSuperTwists; si++)
       {
-        found = true;
-        superIndex.push_back(j);
+        char buf[1000];
+        snprintf(buf, 1000, "Super twist #%d:  [ %9.5f %9.5f %9.5f ]\n", si, superFracs[si][0], superFracs[si][1],
+                 superFracs[si][2]);
+        app_log() << buf;
+        app_log().flush();
       }
     }
-    if (!found)
+  }
+  const int numSuperTwists = superFracs.size();
+
+  { // determine twist_num_
+    std::function find_twist = [&](const TinyVector<double, OHMMS_DIM>& twist) {
+      int twist_num  = -1;
+      PosType gtFrac = FracPart(twist);
+      float eps      = 1e-5;
+      for (int si = 0; si < numSuperTwists; si++)
+      {
+        PosType locDiff = gtFrac - superFracs[si];
+        if (dot(locDiff, locDiff) < eps)
+          twist_num = si;
+      }
+
+      if (twist_num < 0)
+      {
+        char buf[1000];
+        snprintf(buf, 1000,
+                 "AnalyzeTwists2. Input twist [ %9.5f %9.5f %9.5f] not found in the list of super twists above.\n",
+                 twist[0], twist[1], twist[2]);
+        throw UniformCommunicateError(buf);
+      }
+      return twist_num;
+    };
+
+    if (twist_inp[0] > -10 || twist_inp[1] > -10 || twist_inp[2] > -10)
```

Comment:
Why was `-10` chosen here?

Reply:
Good catch. I am not for arbitrary constants. We should use clearer logic.

(I also think we should avoid the arbitrary sized buffers elsewhere in this code. It is not a good code pattern and one we should get away from.)

Reply:
Just added comments in source and commit log.

Reply:
I have replaced -10 with constexpr variables.